lib/msf/core/db_manager/host.rb

```diff
@@ -73,14 +88,35 @@ def add_host_tag(opts)
   # ATM it will delete the tag from the tag table, not the host<->tag link
   def delete_host_tag(opts)
     wspace = Msf::Util::DBManager.process_opts_workspace(opts, framework)
-
-    host_id = opts[:id]
     tag_name = opts[:tag_name]
     tag_ids = []
 
-    host = wspace.hosts.find(host_id)
-    if host
-      found_tags = Mdm::Tag.joins(:hosts).where("hosts.workspace_id = ? and hosts.id = ? and tags.name = ?", wspace.id, host.id, tag_name)
+    # If the command line included an address or address range then use this.
+    # Otherwise delete all entries that match the given tag.
+    if opts[:address]
+      # Find the host entry in the current workspace by searching on
+      # the IP address we are currently processing, then save the database
+      # entry into the "host" variable.
+      # @type [Mdm::Host]
+      host = wspace.hosts.find_by_address(opts[:address])
+      if host
+        found_tags = Mdm::Tag.joins(:hosts).where("hosts.workspace_id = ? and hosts.id = ? and tags.name = ?", wspace.id, host.id, tag_name)
+        found_tags.each do |t|
+          tag_ids << t.id
+        end
+
+        deleted_tags = []
+
+        tag_ids.each do |id|
+          tag = Mdm::Tag.find_by_id(id)
+          deleted_tags << tag
+          tag.destroy
+        end
+
+        return deleted_tags
+      end
+    else
```

Comment:
This feels like dangerous fallback behavior to me; "If I don't find the host you wanted, i'll just delete all the tags"

Reply:
General idea where we modify the logic in the calling scope instead

```ruby
 def delete_host_tag(rws, tag_name)
    opts = Hash.new()
    opts[:workspace] = framework.db.workspace
    opts[:tag_name] = tag_name

    if rws == [nil]
      rws = find_the_hosts_to_delete_the_tags_for
    end
    
    rws.each do |rw|
      rw.each do |ip|
        opts[:address] = ip
        unless framework.db.delete_host_tag(opts)
          print_error("Host #{ip} could not be found.")
        end
      end
    end
  end

```

Reply:
Yeah good point, I'm going to remove this else statement since as you correctly pointed out this is highly likely to lead to inadvertent data loss if a host can't be found. Thanks for pointing this out to me!

Reply:
Should now be handled with 134fd720f006d5d43fad127f0efeb0429a150166, let me know if this works for you @adfoster-r7